src/orderbook.ts

```diff
@@ -122,11 +237,25 @@ function invertPricePoints(prices: Map<number, Offer>) {
   return new Map(
     Array.from(prices.entries()).map(([_, offer]) => {
       const inverted_price = offer.price.inverted();
-      const inverted_volume = offer.volume * offer.price.toNumber();
+      const inverted_volume = offer.volume.mul(offer.price);
       return [
         inverted_price.toNumber(),
         new Offer(inverted_price, inverted_volume)
       ];
     })
   );
 }
+
+// https://github.com/AllAlgorithms/typescript/blob/master/math/gcd/gcd.ts
+function gcd(num1: number, num2: number): number {
+  if (num1 === 0 || num2 === 0) {
+    return 0;
+  }
+  if (num1 === num2) {
+    return num1;
+  }
+  if (num1 > num2) {
+    return gcd(num1 - num2, num2);
+  }
+  return gcd(num1, num2 - num1);
```

Comment:
:crossed_fingers: V8 has TCO

Reply:
Doesn't seem so: https://bugs.chromium.org/p/v8/issues/detail?id=4698

Yet, I don't think this will be our bottleneck when at the same time we are querying data from the blockchain

Reply:
It really shouldn't matter at all. I studied embedded programming so when I see recursion a little alarm goes off in my brain.